server/sonar-server/src/main/java/org/sonar/server/projectlink/ws/CreateAction.java

```diff
@@ -0,0 +1,160 @@
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION_CREATE)
+      .setDescription("Create a new project link.<br>" +
+        "Requires 'Administer' permission on the specified project, " +
+        "or global 'Administer' permission.")
+      .setHandler(this)
+      .setPost(true)
+      .setResponseExample(getClass().getResource("create-example.json"))
+      .setSince("6.1");
+
+    action.createParam(PARAM_PROJECT_ID)
+      .setDescription("Project id")
+      .setExampleValue(UUID_EXAMPLE_01);
+
+    action.createParam(PARAM_PROJECT_KEY)
+      .setDescription("Project key")
+      .setExampleValue(KEY_PROJECT_EXAMPLE_001);
+
+    action.createParam(PARAM_NAME)
+      .setRequired(true)
+      .setDescription("Link name")
+      .setExampleValue("Custom");
+
+    action.createParam(PARAM_URL)
+      .setRequired(true)
+      .setDescription("Link url")
+      .setExampleValue("http://example.com");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    CreateWsRequest searchWsRequest = toCreateWsRequest(request);
+    CreateWsResponse createWsResponse = doHandle(searchWsRequest);
+    writeProtobuf(createWsResponse, request, response);
+  }
+
+  private CreateWsResponse doHandle(CreateWsRequest createWsRequest) {
+    String name = createWsRequest.getName();
+    String url = createWsRequest.getUrl();
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      ComponentDto component = getComponentByUuidOrKey(dbSession, createWsRequest);
+      checkPermissions(component.uuid());
+
+      ComponentLinkDto link = new ComponentLinkDto()
+        .setComponentUuid(component.uuid())
+        .setName(name)
+        .setHref(url)
+        .setType(nameToType(name));
+      dbClient.componentLinkDao().insert(dbSession, link);
+      dbSession.commit();
+      return buildResponse(link);
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private static CreateWsResponse buildResponse(ComponentLinkDto link) {
+    CreateWsResponse.Builder response = CreateWsResponse.newBuilder();
+
+    WsProjectLinks.Link.Builder linkBuilder = WsProjectLinks.Link.newBuilder()
+      .setId(String.valueOf(link.getId()))
+      .setName(link.getName())
+      .setUrl(link.getHref());
+
+    String type = link.getType();
+    if (type != null) {
+      linkBuilder.setType(type);
+    }
+
+    response.setLink(linkBuilder);
+
+    return response.build();
+  }
+
+  private ComponentDto getComponentByUuidOrKey(DbSession dbSession, CreateWsRequest request) {
+    return componentFinder.getByUuidOrKey(
+      dbSession,
+      request.getProjectId(),
+      request.getProjectKey(),
+      ComponentFinder.ParamNames.PROJECT_ID_AND_KEY);
+  }
+
+  private void checkPermissions(String projectUuid) {
+    if (!userSession.hasComponentUuidPermission(UserRole.ADMIN, projectUuid)) {
```

Comment:
use `userSession.checkComponentUuidPermission(UserRole.ADMIN, projectUuid)` or even inline the method